internal/pkg/container/syncuids.go

```diff
@@ -16,243 +16,300 @@ import (
-	name string
-	uid  int
-	gid  int
-}
 
-/*
-sync the uids,gids from the host to the container
-*/
 func SyncUids(containerName string, showOnly bool) error {
-	var userDb []completeUserInfo
-	passwdName := "/etc/passwd"
-	groupName := "/etc/group"
-
-	// populate db with users from the host
-	hostUsers, err := createPasswdMap(passwdName)
-	if err != nil {
-		wwlog.Error("Could not open "+passwdName)
-		return err
-	}
-	for _, hostUser := range hostUsers {
-		userDb = append(userDb, completeUserInfo{Name: hostUser.name,
-			UidHost: hostUser.uid, GidHost: hostUser.gid, UidCont: -1, GidCont: -1})
-	}
+	passwdPath := "/etc/passwd"
+	groupPath := "/etc/group"
+	containerPath := RootFsDir(containerName)
+	containerPasswdPath := path.Join(containerPath, passwdPath)
+	containerGroupPath := path.Join(containerPath, groupPath)
 
-	// merge container users into db and track users that are only in the
-	// container
-	fullPath := RootFsDir(containerName)
-	containerUsers, err := createPasswdMap(path.Join(fullPath, passwdName))
-	if err != nil {
-		wwlog.Error("Could not open "+path.Join(fullPath, passwdName))
-		return err
-	}
-	var userOnlyCont []string
-	for _, containerUser := range containerUsers {
-		foundUser := false
-		for idxHost, user := range userDb {
-			if containerUser.name == user.Name {
-				foundUser = true
-				(&userDb[idxHost]).UidCont = containerUser.uid
-				(&userDb[idxHost]).GidCont = containerUser.gid
-			}
-		}
-		if !foundUser {
-			userDb = append(userDb, completeUserInfo{Name: containerUser.name,
-				UidHost: -1, GidHost: -1, UidCont: containerUser.uid, GidCont: containerUser.gid})
-			wwlog.Warn("user: %s:%v:%v not present on host", containerUser.name, containerUser.uid, containerUser.gid)
-			userOnlyCont = append(userOnlyCont, containerUser.name)
+	passwdSync := make(syncDB)
+	if err := passwdSync.read(passwdPath, false); err != nil { return err }
+	if err := passwdSync.read(containerPasswdPath, true); err != nil { return err }
+	if err := passwdSync.checkConflicts(); err != nil { return err }
+	if err := passwdSync.findUserFiles(containerPath); err != nil { return err }
+
+	groupSync := make(syncDB)
+	if err := groupSync.read(groupPath, false); err != nil { return err }
+	if err := groupSync.read(containerGroupPath, true); err != nil { return err }
+	if err := groupSync.checkConflicts(); err != nil { return err }
+	if err := groupSync.findGroupFiles(containerPath); err != nil { return err }
+
+	passwdSync.log("passwd")
+	groupSync.log("group")
+
+	if showOnly {
+		if passwdSync.needsSync() || groupSync.needsSync() {
+			wwlog.Info("uid/gid not synced: run `wwctl container syncuser --write %s`", containerName)
+		} else {
+			wwlog.Info("uid/gid already synced")
 		}
+	} else {
+		if err := passwdSync.chownUserFiles(); err != nil { return err }
+		if err := groupSync.chownGroupFiles(); err != nil { return err }
+
+		if err := passwdSync.update(containerPasswdPath, passwdPath); err != nil { return err }
```

Comment:
we ca add check here
if passwdSync.needsSync(){
    passwdSync.chownUserFiles();
}
if groupSync.needsSync(){
    ....
}